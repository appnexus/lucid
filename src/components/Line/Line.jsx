import _ from 'lodash';
import React from 'react';
import PropTypes from 'prop-types';
import { lucidClassNames } from '../../util/style-helpers';
import {
	createClass as createReactClass,
	omitProps,
} from '../../util/component-types';
import * as chartConstants from '../../constants/charts';

const cx = lucidClassNames.bind('&-Line');

const { string, object, bool } = PropTypes;

/**
 * {"categories": ["visualizations", "geoms"]}
 *
 * *For use within an `svg`*
 *
 * Lines are typically used for line charts and are pretty much a thin wrapper
 * around svg paths.
 *
 */
const Line = createReactClass({
	displayName: 'Line',

	propTypes: {
		/**
		 * Passed through to the root element.
		 */
		style: object,
		/**
		 * Appended to the component-specific class names set on the root element.
		 */
		className: string,
		/**
		 * The path for the line.
		 */
		d: string,
		/**
		 * Strings should match an existing color class unless they start with a
		 * '#' for specific colors. E.g.:
		 *
		 * - `COLOR_0`
		 * - `COLOR_GOOD`
		 * - `'#123abc'`
		 */
		color: string,
		/**
		 * Display a dotted line.
		 */
		isDotted: bool,
	},

	getDefaultProps() {
		return {
			color: chartConstants.COLOR_0,
			isDotted: false,
		};
	},

	render() {
		const {
			className,
			color,
			isDotted,
			d,
			style,
			...passThroughs
		} = this.props;

		const isCustomColor = _.startsWith(color, '#');
		const colorStyle = isCustomColor ? { fill: color, stroke: color } : null;

		return (
			<path
				{...omitProps(passThroughs, Line)}
				style={{
					...style,
					...colorStyle,
				}}
				className={cx(className, '&', {
					[`&-${color}`]: !isCustomColor,
					'&-is-dotted': isDotted,
				})}
				d={d}
			/>
		);
	},
});

export default Line;
