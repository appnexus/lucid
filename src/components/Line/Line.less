@import (reference) '../../styles/variables.less';
@import (reference) '../../styles/mixins.less';

.lucid-Line {
	.chart-color-loop(@stroke: true, @fill: true);
	stroke-width: 2;

	&-is-dotted {
		stroke-dasharray: 6;
	}
}
