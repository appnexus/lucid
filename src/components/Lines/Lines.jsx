import _ from 'lodash';
import React from 'react';
import PropTypes from 'prop-types';
import { lucidClassNames } from '../../util/style-helpers';
import {
	createClass as createReactClass,
	omitProps,
} from '../../util/component-types';
import { groupByFields } from '../../util/chart-helpers';
import * as d3Shape from 'd3-shape';
import * as chartConstants from '../../constants/charts';

import Line from '../Line/Line';

const cx = lucidClassNames.bind('&-Lines');

const { arrayOf, func, number, object, bool, string } = PropTypes;

/**
 * {"categories": ["visualizations", "chart primitives"], "madeFrom": ["Line"]}
 *
 * *For use within an `svg`*
 *
 * Lines are typically used to represent continuous data and can be stacked.
 */
const Lines = createReactClass({
	displayName: 'Lines',

	propTypes: {
		/**
		 * Appended to the component-specific class names set on the root element.
		 */
		className: string,
		/**
		 * Top
		 */
		top: number,
		/**
		 * Left
		 */
		left: number,
		/**
		 * Takes one of the palettes exported from `lucid.chartConstants`.
		 * Available palettes:
		 *
		 * - `PALETTE_6` (default)
		 * - `PALETTE_30`
		 * - `PALETTE_MONOCHROME_0_5`
		 * - `PALETTE_MONOCHROME_1_5`
		 * - `PALETTE_MONOCHROME_2_5`
		 * - `PALETTE_MONOCHROME_3_5`
		 * - `PALETTE_MONOCHROME_4_5`
		 * - `PALETTE_MONOCHROME_5_5`
		 *
		 */
		palette: arrayOf(string),
		/**
		 * You can pass in an object if you want to map fields to
		 * `lucid.chartConstants` or custom colors:
		 *
		 *     {
		 *       'imps': COLOR_0,
		 *       'rev': COLOR_3,
		 *       'clicks': '#abc123',
		 *     }
		 */
		colorMap: object,
		/**
		 * De-normalized data, e.g.
		 *
		 *
		 *     [
		 *       { x: 'one'   , y: 1 },
		 *       { x: 'two'   , y: 2 },
		 *       { x: 'three' , y: 2 },
		 *       { x: 'four'  , y: 3 },
		 *       { x: 'five'  , y: 4 },
		 *     ]
		 *
		 *
		 * Or (be sure to set `yFields` to `['y0', 'y1', 'y2', 'y3']`)
		 *
		 *     [
		 *       { x: 'one'   , y0: 1  , y1: 2 , y2: 3 , y3: 5 },
		 *       { x: 'two'   , y0: 2  , y1: 3 , y2: 4 , y3: 6 },
		 *       { x: 'three' , y0: 2  , y1: 4 , y2: 5 , y3: 6 },
		 *       { x: 'four'  , y0: 3  , y1: 6 , y2: 7 , y3: 7 },
		 *       { x: 'five'  , y0: 4  , y1: 8 , y2: 9 , y3: 8 },
		 *       { x: 'six'   , y0: 20 , y1: 8 , y2: 9 , y3: 1 },
		 *     ]
		 *
		 */
		data: arrayOf(object).isRequired,
		/**
		 * The scale for the x axis. Must be a d3 scale. Lucid exposes the
		 * `lucid.d3Scale` library for use here.
		 */
		xScale: func.isRequired,
		/**
		 * The scale for the y axis. Must be a d3 scale. Lucid exposes the
		 * `lucid.d3Scale` library for use here.
		 */
		yScale: func.isRequired,
		/**
		 * Typically this number can be derived from the yScale. However when we're
		 * `isStacked` we need to calculate a new domain for the yScale based on
		 * the sum of the data. If you need explicit control of the y max when
		 * stacking, pass it in here.
		 */
		yStackedMax: number,
		/**
		 * The field we should look up your x data by.
		 */
		xField: string,
		/**
		 * The field(s) we should look up your y data by. Each entry represents a
		 * series. Your actual y data should be numeric.
		 */
		yFields: arrayOf(string),
		/**
		 * This will stack the data instead of grouping it. In order to stack the
		 * data we have to calculate a new domain for the y scale that is based on
		 * the `sum` of the data.
		 */
		isStacked: bool,
		/**
		 * Sometimes you might not want the colors to start rotating at the blue
		 * color, this number will be added the line index in determining which
		 * color the lines are.
		 */
		colorOffset: number,
	},

	getDefaultProps() {
		return {
			xField: 'x',
			yFields: ['y'],
			isStacked: false,
			colorOffset: 0,
			palette: chartConstants.PALETTE_6,
		};
	},

	render() {
		const {
			className,
			data,
			isStacked,
			palette,
			colorMap,
			colorOffset,
			xScale,
			xField,
			yFields,
			yScale: yScaleOriginal,
			yStackedMax,
			...passThroughs
		} = this.props;

		// Copy the original so we can mutate it
		const yScale = yScaleOriginal.copy();

		// If we are stacked, we need to calculate a new domain based on the sum of
		// the various series' y data. One row per series.
		const transformedData = isStacked
			? d3Shape.stack().keys(yFields)(data)
			: groupByFields(data, yFields);

		const area = isStacked
			? d3Shape
					.area()
					.defined(a => _.isFinite(a[0]) && _.isFinite(a[1]))
					.x((a, i) => xScale(data[i][xField]))
					.y0(a => yScale(a[1]))
					.y1(a => yScale(a[0]))
			: d3Shape
					.area()
					.defined(a => _.isFinite(a) || _.isDate(a))
					.x((a, i) => xScale(data[i][xField]))
					.y(a => yScale(a));

		// If we are stacked, we need to calculate a new domain based on the sum of
		// the various group's y data
		if (isStacked) {
			yScale.domain([
				yScale.domain()[0], // only stacks well if this is `0`
				yStackedMax || _.max(_.flatten(_.last(transformedData))),
			]);
		}

		return (
			<g {...omitProps(passThroughs, Lines)} className={cx(className, '&')}>
				{_.map(transformedData, (d, dIndex) => (
					<g key={dIndex}>
						<Line
							color={_.get(
								colorMap,
								yFields[dIndex],
								palette[(dIndex + colorOffset) % palette.length]
							)}
							d={area(d)}
						/>
					</g>
				))}
			</g>
		);
	},
});

export default Lines;
