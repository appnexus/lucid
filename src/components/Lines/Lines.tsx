import _ from 'lodash';
import React from 'react';
import PropTypes from 'prop-types';
import * as d3Shape from 'd3-shape';
import * as d3Scale from 'd3-scale';

import { lucidClassNames } from '../../util/style-helpers';
import { StandardProps } from '../../util/component-types';
import { groupByFields } from '../../util/chart-helpers';
import * as chartConstants from '../../constants/charts';
import Line from '../Line/Line';

const cx = lucidClassNames.bind('&-Lines');

const { arrayOf, func, number, object, bool, string } = PropTypes;

const isUniform = (array: any[]): boolean =>
	_.every(array, (val): boolean => val === _.first(array));

export interface ILinesProps
	extends StandardProps,
		React.SVGProps<SVGGElement> {
	/** Top */
	top?: number;

	/** Left */
	left?: number;

	/** 	Takes one of the palettes exported from \`lucid.chartConstants\`.
	Available palettes:

	- \`PALETTE_7\` (default)
	- \`PALETTE_30\`
	- \`PALETTE_MONOCHROME_0_5\`
	- \`PALETTE_MONOCHROME_1_5\`
	- \`PALETTE_MONOCHROME_2_5\`
	- \`PALETTE_MONOCHROME_3_5\`
	- \`PALETTE_MONOCHROME_4_5\`
	- \`PALETTE_MONOCHROME_5_5\`
	- \`PALETTE_MONOCHROME_6_5\` */
	palette: string[];

	/** You can pass in an object if you want to map fields to
	\`lucid.chartConstants\` or custom colors:

		{
			'imps': COLOR_0,
			'rev': COLOR_3,
			'clicks': '#abc123',
		} */
	colorMap?: object;

	/** De-normalized data, e.g.

		[
			{ x: 'one'   , y: 1 },
			{ x: 'two'   , y: 2 },
			{ x: 'three' , y: 2 },
			{ x: 'four'  , y: 3 },
			{ x: 'five'  , y: 4 },
		]

	Or (be sure to set \`yFields\` to \`['y0', 'y1', 'y2', 'y3']\`)

		[
			{ x: 'one'   , y0: 1  , y1: 2 , y2: 3 , y3: 5 },
			{ x: 'two'   , y0: 2  , y1: 3 , y2: 4 , y3: 6 },
			{ x: 'three' , y0: 2  , y1: 4 , y2: 5 , y3: 6 },
			{ x: 'four'  , y0: 3  , y1: 6 , y2: 7 , y3: 7 },
			{ x: 'five'  , y0: 4  , y1: 8 , y2: 9 , y3: 8 },
			{ x: 'six'   , y0: 20 , y1: 8 , y2: 9 , y3: 1 },
		] */

	data: Array<{ [key: string]: any }>;

	//TODO: xScale can support several different types of scales, maybe all types; we need to enumerate what it acccepts and encode it into the type
	// Having a number | string union type gets converted by TypeScript to ReactText; It may be related to the LibraryManagedAttributes
	/** The scale for the x axis. Must be a d3 scale. Lucid exposes the
		`lucid.d3Scale` library for use here. */
	xScale:
		| d3Scale.ScaleBand<string>
		| d3Scale.ScalePoint<string>
		| d3Scale.ScaleTime<number, number>;

	/** The scale for the y axis. Must be a d3 scale. Lucid exposes the
		`lucid.d3Scale` library for use here. */
	yScale:
		| d3Scale.ScaleContinuousNumeric<number, number>
		| d3Scale.ScaleBand<number>
		| d3Scale.ScalePoint<number>
		| d3Scale.ScaleLinear<number, number>;

	/** Typically this number can be derived from the yScale. However when we're
		\`isStacked\` we need to calculate a new domain for the yScale based on
		the sum of the data. If you need explicit control of the y max when
		stacking, pass it in here. */
	yStackedMax?: number | object;

	/** The field we should look up your x data by. */
	xField: string;

	/** The field(s) we should look up your y data by. Each entry represents a
		series. Your actual y data should be numeric. */
	yFields: string[];

	/** This will stack the data instead of grouping it. In order to stack the
		data we have to calculate a new domain for the y scale that is based on
		the \`sum\` of the data. */
	isStacked: boolean;

	/** Sometimes you might not want the colors to start rotating at the blue
		color, this number will be added the line index in determining which
		color the lines are. */
	colorOffset: number;
}

const defaultProps = {
	xField: 'x',
	yFields: ['y'],
	isStacked: false,
	colorOffset: 0,
	palette: chartConstants.PALETTE_7,
};

export const Lines = (props: ILinesProps): React.ReactElement => {
	const {
		className,
		data,
		isStacked,
		palette,
		colorMap,
		colorOffset,
		xScale,
		xField,
		yFields,
		yScale: yScaleOriginal,
		yStackedMax,
		...passThroughs
	} = props;

	// Copy the original so we can mutate it
	const yScale = yScaleOriginal.copy();

	// If we are stacked, we need to calculate a new domain based on the sum of
	// the various series' y data. One row per series.
	const transformedData = isStacked
		? d3Shape.stack().keys(yFields)(data as Array<{ [key: string]: number }>)
		: groupByFields(data, yFields);

	const stackedArea = d3Shape
		.area<[number, number]>()
		.defined((a): boolean => _.isFinite(a[0]) && _.isFinite(a[1]))
		.x((a, i): number => xScale(data[i][xField]) as number)
		.y0((a): number => yScale(a[1]) as number)
		.y1((a): number => yScale(a[0]) as number);

	const area = d3Shape
		.area<number>()
		.defined((a): boolean => _.isFinite(a) || _.isDate(a))
		.x((a, i): number => xScale(data[i][xField]) as number)
		.y((a, i): number => yScale(a) as number);

	// If we are stacked, we need to calculate a new domain based on the sum of
	// the various group's y data
	if (isStacked) {
		yScale.domain([
			yScale.domain()[0], // only stacks well if this is `0`
			yStackedMax || _.max(_.flatten(_.last(transformedData))),
		]);
	}

	return (
		<g {...passThroughs} className={cx(className, '&')}>
			{_.map(
				transformedData,
				(d, dIndex): React.ReactElement => (
					<g key={dIndex}>
						<Line
							color={_.get(
								colorMap,
								yFields[dIndex],
								palette[(dIndex + colorOffset) % palette.length]
							)}
							d={(isStacked ? stackedArea(d) : area(d)) as string}
							// Fixes a bug in chrome by forcing a reflow of the element (ANXR-1405)
							style={isUniform(d) ? { transform: 'scaleX(0.999)' } : undefined}
						/>
					</g>
				)
			)}
		</g>
	);
};

Lines.defaultProps = defaultProps;
Lines.displayName = 'Lines';
Lines.peek = {
	description: `*For use within an \`svg\`*. A \`Line\` is typically used to represent continuous data and can be stacked.`,
	categories: ['visualizations', 'chart primitives'],
	madeFrom: ['Line'],
};
Lines.propTypes = {
	/**
		Appended to the component-specific class names set on the root element.
	*/
	className: string,

	/**
		Top
	*/
	top: number,

	/**
		Left
	*/
	left: number,

	/**
		Takes one of the palettes exported from \`lucid.chartConstants\`.
		Available palettes:

		- \`PALETTE_7\` (default)
		- \`PALETTE_30\`
		- \`PALETTE_MONOCHROME_0_5\`
		- \`PALETTE_MONOCHROME_1_5\`
		- \`PALETTE_MONOCHROME_2_5\`
		- \`PALETTE_MONOCHROME_3_5\`
		- \`PALETTE_MONOCHROME_4_5\`
		- \`PALETTE_MONOCHROME_5_5\`
		- \`PALETTE_MONOCHROME_6_5\`
	*/
	palette: arrayOf(string),

	/**
		You can pass in an object if you want to map fields to
		\`lucid.chartConstants\` or custom colors:

			{
				'imps': COLOR_0,
				'rev': COLOR_3,
				'clicks': '#abc123',
			}
	*/
	colorMap: object,

	/**
		De-normalized data, e.g.

			[
				{ x: 'one'   , y: 1 },
				{ x: 'two'   , y: 2 },
				{ x: 'three' , y: 2 },
				{ x: 'four'  , y: 3 },
				{ x: 'five'  , y: 4 },
			]


		Or (be sure to set \`yFields\` to \`['y0', 'y1', 'y2', 'y3']\`)

			[
				{ x: 'one'   , y0: 1  , y1: 2 , y2: 3 , y3: 5 },
				{ x: 'two'   , y0: 2  , y1: 3 , y2: 4 , y3: 6 },
				{ x: 'three' , y0: 2  , y1: 4 , y2: 5 , y3: 6 },
				{ x: 'four'  , y0: 3  , y1: 6 , y2: 7 , y3: 7 },
				{ x: 'five'  , y0: 4  , y1: 8 , y2: 9 , y3: 8 },
				{ x: 'six'   , y0: 20 , y1: 8 , y2: 9 , y3: 1 },
			]
	*/
	data: arrayOf(object).isRequired,

	/**
		The scale for the x axis. Must be a d3 scale. Lucid exposes the
		\`lucid.d3Scale\` library for use here.
	*/
	xScale: func.isRequired,

	/**
		The scale for the y axis. Must be a d3 scale. Lucid exposes the
		\`lucid.d3Scale\` library for use here.
	*/
	yScale: func.isRequired,

	/**
		Typically this number can be derived from the yScale. However when we're
		\`isStacked\` we need to calculate a new domain for the yScale based on
		the sum of the data. If you need explicit control of the y max when
		stacking, pass it in here.
	*/
	yStackedMax: number,

	/**
		The field we should look up your x data by.
	*/
	xField: string,

	/**
		The field(s) we should look up your y data by. Each entry represents a
		series. Your actual y data should be numeric.
	*/
	yFields: arrayOf(string),

	/**
		This will stack the data instead of grouping it. In order to stack the
		data we have to calculate a new domain for the y scale that is based on
		the \`sum\` of the data.
	*/
	isStacked: bool,

	/**
		Sometimes you might not want the colors to start rotating at the blue
		color, this number will be added the line index in determining which
		color the lines are.
	*/
	colorOffset: number,
};

export default Lines;
