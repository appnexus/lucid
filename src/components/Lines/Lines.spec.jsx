/* eslint-disable comma-spacing */

import React from 'react';
import assert from 'assert';
import * as d3Scale from 'd3-scale';
import { common } from '../../util/generic-tests';
import { shallow } from 'enzyme';

import Lines from './Lines';
import Line from '../Line/Line';

const defaultData = [
	{ x: 0, y: 10, y2: 20 },
	{ x: 5, y: 55, y2: 35 },
	{ x: 10, y: 100, y2: 0 },
];

// The scales are functions to guard against inadvertent scale mutation
const defaultXScale = d3Scale
	.scaleLinear()
	.domain([0, 10])
	.range([0, 100]);

const defaultYScale = d3Scale
	.scaleLinear()
	.domain([0, 100])
	.range([1000, 0]);

describe('Lines', () => {
	common(Lines, {
		exemptFunctionProps: ['xScale', 'yScale'],
		getDefaultProps: () => ({
			data: [{ x: new Date('2015-01-01T00:00:00Z'), y: 1 }],
			xScale: d3Scale.scaleTime(),
			yScale: d3Scale.scaleLinear(),
		}),
	});

	describe('props', () => {
		describe('palette', () => {
			it('should kick in when theres no colorMap', () => {
				const palette = ['bling', 'blang'];
				const wrapper = shallow(
					<Lines
						data={[{ x: 1, rev: 1, imp: 2, click: 3 }]}
						yFields={['rev', 'imp', 'click']}
						xScale={defaultXScale}
						yScale={defaultYScale}
						palette={palette}
					/>
				);

				assert.equal(
					wrapper
						.find(Line)
						.at(0)
						.prop('color'),
					'bling'
				);
				assert.equal(
					wrapper
						.find(Line)
						.at(1)
						.prop('color'),
					'blang'
				);
				assert.equal(
					wrapper
						.find(Line)
						.at(2)
						.prop('color'),
					'bling'
				);
			});

			it('should be beat when theres a colorMap', () => {
				const palette = ['bling', 'blang'];
				const wrapper = shallow(
					<Lines
						data={[{ x: 1, rev: 1, imp: 2, click: 3 }]}
						yFields={['rev', 'imp', 'click']}
						xScale={defaultXScale}
						yScale={defaultYScale}
						palette={palette}
						colorMap={{
							rev: 'wat',
						}}
					/>
				);

				assert.equal(
					wrapper
						.find(Line)
						.at(0)
						.prop('color'),
					'wat'
				);
				assert.equal(
					wrapper
						.find(Line)
						.at(1)
						.prop('color'),
					'blang'
				);
				assert.equal(
					wrapper
						.find(Line)
						.at(2)
						.prop('color'),
					'bling'
				);
			});
		});

		describe('colorMap', () => {
			it('should pass custom colors through to Line', () => {
				const wrapper = shallow(
					<Lines
						data={defaultData}
						xScale={defaultXScale}
						yScale={defaultYScale}
						colorMap={{
							y: '#ABC333',
						}}
					/>
				);

				assert.equal(
					wrapper
						.find(Line)
						.at(0)
						.prop('color'),
					'#ABC333'
				);
			});
		});

		describe('data', () => {
			it('should handle a basic data set', () => {
				const wrapper = shallow(
					<Lines
						data={defaultData}
						xScale={defaultXScale}
						yScale={defaultYScale}
					/>
				);

				assert.equal(wrapper.find(Line).length, 1);
			});
		});

		describe('xScale', () => {
			it('should work with linear scales', () => {
				const wrapper = shallow(
					<Lines
						data={defaultData}
						xScale={defaultXScale}
						yScale={defaultYScale}
					/>
				);

				assert.equal(wrapper.find(Line).length, 1, 'wrong number of Line');
				assert.equal(
					wrapper
						.find(Line)
						.at(0)
						.prop('d'),
					'M0,900L50,449.99999999999994L100,0L100,0L50,449.99999999999994L0,900Z'
				);
			});

			it('should work with time scales', () => {
				const xScale = d3Scale
					.scaleTime()
					.domain([
						new Date('2015-01-01T00:00:00Z'),
						new Date('2015-01-02T00:00:00Z'),
					])
					.range([0, 500]);
				const wrapper = shallow(
					<Lines
						data={[
							{ x: new Date('2015-01-01T00:00:00Z'), y: 50 },
							{ x: new Date('2015-01-02T00:00:00Z'), y: 60 },
						]}
						xScale={xScale}
						yScale={defaultYScale}
					/>
				);

				assert.equal(wrapper.find(Line).length, 1, 'wrong number of Line');
				assert.equal(
					wrapper
						.find(Line)
						.at(0)
						.prop('d'),
					'M0,500L500,400L500,400L0,500Z'
				);
			});
		});

		describe('yScale', () => {
			it('should work with time scales', () => {
				const yScale = d3Scale
					.scaleTime()
					.domain([
						new Date('2015-01-01T00:00:00Z'),
						new Date('2015-01-02T00:00:00Z'),
					])
					.range([0, 500]);
				const wrapper = shallow(
					<Lines
						data={[
							{ y: new Date('2015-01-01T00:00:00Z'), x: 5 },
							{ y: new Date('2015-01-02T00:00:00Z'), x: 6 },
						]}
						xScale={defaultXScale}
						yScale={yScale}
					/>
				);

				assert.equal(wrapper.find(Line).length, 1, 'wrong number of Line');
				assert.equal(
					wrapper
						.find(Line)
						.at(0)
						.prop('d'),
					'M50,0L60,500L60,500L50,0Z'
				);
			});
		});

		describe('xField', () => {
			it('should pickup alternate xFields', () => {
				const wrapper = shallow(
					<Lines
						data={[{ axe: 1, y: 3 }, { axe: 2, y: 2 }, { axe: 3, y: 1 }]}
						xField='axe'
						xScale={defaultXScale}
						yScale={defaultYScale}
					/>
				);

				assert.equal(wrapper.find(Line).length, 1, 'wrong number of Line');
				assert.equal(
					wrapper
						.find(Line)
						.at(0)
						.prop('d'),
					'M10,970L20,980L30,990L30,990L20,980L10,970Z'
				);
			});
		});

		describe('yFields', () => {
			it('should pickup an alternate field', () => {
				const wrapper = shallow(
					<Lines
						data={defaultData}
						yFields={['y2']}
						xScale={defaultXScale}
						yScale={defaultYScale}
					/>
				);

				assert.equal(
					wrapper
						.find(Line)
						.at(0)
						.prop('d'),
					'M0,800L50,650L100,1000L100,1000L50,650L0,800Z'
				);
			});

			it('should handle multiple fields', () => {
				const wrapper = shallow(
					<Lines
						data={defaultData}
						yFields={['y', 'y2']}
						xScale={defaultXScale}
						yScale={defaultYScale}
					/>
				);

				assert.equal(
					wrapper
						.find(Line)
						.at(0)
						.prop('d'),
					'M0,900L50,449.99999999999994L100,0L100,0L50,449.99999999999994L0,900Z'
				);
				assert.equal(
					wrapper
						.find(Line)
						.at(1)
						.prop('d'),
					'M0,800L50,650L100,1000L100,1000L50,650L0,800Z'
				);
			});
		});

		describe('isStacked', () => {
			it('should stack', () => {
				const wrapper = shallow(
					<Lines
						data={defaultData}
						yFields={['y', 'y2']}
						xScale={defaultXScale}
						yScale={defaultYScale}
						isStacked
					/>
				);

				assert.equal(
					wrapper
						.find(Line)
						.at(0)
						.prop('d'),
					'M0,1000L50,1000L100,1000L100,0L50,449.99999999999994L0,900Z'
				);
				assert.equal(
					wrapper
						.find(Line)
						.at(1)
						.prop('d'),
					'M0,900L50,449.99999999999994L100,0L100,0L50,99.99999999999997L0,700Z'
				);
			});
		});

		describe('colorOffset', () => {
			it('should correctly offset the palette colors', () => {
				const palette = ['bing', 'bong'];
				const wrapper = shallow(
					<Lines
						data={[{ x: 1, rev: 1, imp: 2, click: 3 }]}
						yFields={['rev', 'imp', 'click']}
						xScale={defaultXScale}
						yScale={defaultYScale}
						palette={palette}
						colorOffset={1}
					/>
				);

				assert.equal(
					wrapper
						.find(Line)
						.at(0)
						.prop('color'),
					'bong'
				);
				assert.equal(
					wrapper
						.find(Line)
						.at(1)
						.prop('color'),
					'bing'
				);
				assert.equal(
					wrapper
						.find(Line)
						.at(2)
						.prop('color'),
					'bong'
				);
			});
		});
	});
});
