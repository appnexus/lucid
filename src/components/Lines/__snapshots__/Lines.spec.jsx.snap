// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lines [common] example testing should match snapshot(s) for basic 1`] = `
<g
  className="lucid-Lines"
>
  <g
    key="0"
  >
    <Line
      color="color-chart-3"
      d="M0,0L160,360L320,360L480,340L640,320L800,0L800,0L640,320L480,340L320,360L160,360L0,0Z"
      isDotted={false}
    />
  </g>
  <g
    key="1"
  >
    <Line
      color="color-chart-0-darkest"
      d="M0,39.99999999999999L160,340L320,320L480,280L640,240L800,39.99999999999999L800,39.99999999999999L640,240L480,280L320,320L160,340L0,39.99999999999999Z"
      isDotted={false}
    />
  </g>
  <g
    key="2"
  >
    <Line
      color="color-chart-2"
      d="M0,79.99999999999999L160,320L320,300L480,260L640,220.00000000000003L800,79.99999999999999L800,79.99999999999999L640,220.00000000000003L480,260L320,300L160,320L0,79.99999999999999Z"
      isDotted={false}
    />
  </g>
  <g
    key="3"
  >
    <Line
      color="color-chart-4-lightest"
      d="M0,120.00000000000001L160,280L320,280L480,260L640,240L800,120.00000000000001L800,120.00000000000001L640,240L480,260L320,280L160,280L0,120.00000000000001Z"
      isDotted={false}
    />
  </g>
  <g
    key="4"
  >
    <Line
      color="color-chart-5"
      d="M0,160L160,240L320,220.00000000000003L480,260L640,300L800,160L800,160L640,300L480,260L320,220.00000000000003L160,240L0,160Z"
      isDotted={false}
    />
  </g>
  <g
    key="5"
  >
    <Line
      color="color-chart-1-light"
      d="M0,200L160,220.00000000000003L320,200L480,320L640,320L800,200L800,200L640,320L480,320L320,200L160,220.00000000000003L0,200Z"
      isDotted={false}
    />
  </g>
  <g
    key="6"
  >
    <Line
      color="color-chart-6"
      d="M0,240L160,280L320,340L480,360L640,340L800,240L800,240L640,340L480,360L320,340L160,280L0,240Z"
      isDotted={false}
    />
  </g>
</g>
`;

exports[`Lines [common] example testing should match snapshot(s) for basic 2`] = `
<g
  className="lucid-Lines"
>
  <g
    key="0"
  >
    <Line
      color="color-chart-3"
      d="M0,400L160,400L320,400L480,400L640,400L800,400L800,318.3673469387755L640,383.67346938775506L480,387.7551020408163L320,391.83673469387753L160,391.83673469387753L0,318.3673469387755Z"
      isDotted={false}
    />
  </g>
  <g
    key="1"
  >
    <Line
      color="color-chart-0-darkest"
      d="M0,318.3673469387755L160,391.83673469387753L320,391.83673469387753L480,387.7551020408163L640,383.67346938775506L800,318.3673469387755L800,244.89795918367346L640,351.0204081632653L480,363.265306122449L320,375.51020408163265L160,379.5918367346939L0,244.89795918367346Z"
      isDotted={false}
    />
  </g>
  <g
    key="2"
  >
    <Line
      color="color-chart-2"
      d="M0,244.89795918367346L160,379.5918367346939L320,375.51020408163265L480,363.265306122449L640,351.0204081632653L800,244.89795918367346L800,179.5918367346939L640,314.2857142857143L480,334.6938775510204L320,355.1020408163265L160,363.265306122449L0,179.5918367346939Z"
      isDotted={false}
    />
  </g>
  <g
    key="3"
  >
    <Line
      color="color-chart-4-lightest"
      d="M0,179.5918367346939L160,363.265306122449L320,355.1020408163265L480,334.6938775510204L640,314.2857142857143L800,179.5918367346939L800,122.44897959183673L640,281.6326530612245L480,306.1224489795918L320,330.6122448979592L160,338.7755102040816L0,122.44897959183673Z"
      isDotted={false}
    />
  </g>
  <g
    key="4"
  >
    <Line
      color="color-chart-5"
      d="M0,122.44897959183673L160,338.7755102040816L320,330.6122448979592L480,306.1224489795918L640,281.6326530612245L800,122.44897959183673L800,73.46938775510203L640,261.2244897959184L480,277.55102040816325L320,293.87755102040813L160,306.1224489795918L0,73.46938775510203Z"
      isDotted={false}
    />
  </g>
  <g
    key="5"
  >
    <Line
      color="color-chart-1-light"
      d="M0,73.46938775510203L160,306.1224489795918L320,293.87755102040813L480,277.55102040816325L640,261.2244897959184L800,73.46938775510203L800,32.65306122448979L640,244.89795918367346L480,261.2244897959184L320,253.06122448979593L160,269.38775510204084L0,32.65306122448979Z"
      isDotted={false}
    />
  </g>
  <g
    key="6"
  >
    <Line
      color="color-chart-6"
      d="M0,32.65306122448979L160,269.38775510204084L320,253.06122448979593L480,261.2244897959184L640,244.89795918367346L800,32.65306122448979L800,0L640,232.6530612244898L480,253.06122448979593L320,240.81632653061226L160,244.89795918367346L0,0Z"
      isDotted={false}
    />
  </g>
</g>
`;

exports[`Lines [common] example testing should match snapshot(s) for custom-colors 1`] = `
<g
  className="lucid-Lines"
>
  <g
    key="0"
  >
    <Line
      color="color-chart-bad"
      d="M0,380L200,360L400,360L600,340L800,320L1000,0L1000,0L800,320L600,340L400,360L200,360L0,380Z"
      isDotted={false}
    />
  </g>
  <g
    key="1"
  >
    <Line
      color="color-chart-good"
      d="M0,360L200,340L400,320L600,280L800,240L1000,240L1000,240L800,240L600,280L400,320L200,340L0,360Z"
      isDotted={false}
    />
  </g>
  <g
    key="2"
  >
    <Line
      color="#ff8800"
      d="M0,340L200,320L400,300L600,260L800,220.00000000000003L1000,220.00000000000003L1000,220.00000000000003L800,220.00000000000003L600,260L400,300L200,320L0,340Z"
      isDotted={false}
    />
  </g>
  <g
    key="3"
  >
    <Line
      color="#abc123"
      d="M0,300L200,280L400,280L600,260L800,240L1000,380L1000,380L800,240L600,260L400,280L200,280L0,300Z"
      isDotted={false}
    />
  </g>
</g>
`;
